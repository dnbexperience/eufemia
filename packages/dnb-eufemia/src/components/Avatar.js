/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index avatar to autogenerate all the components and extensions
 * Used by "prepareAvatar"
 */

import Avatar from './avatar/Avatar'
export * from './avatar/Avatar'
export default Avatar
