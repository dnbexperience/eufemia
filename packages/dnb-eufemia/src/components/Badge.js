/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index bade to autogenerate all the components and extensions
 * Used by "prepareBadge"
 */

import Badge from './badge/Badge'
export * from './badge/Badge'
export default Badge
