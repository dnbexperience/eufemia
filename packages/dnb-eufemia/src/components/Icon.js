/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index icon to autogenerate all the components and patterns
 * Used by "prepareIcons"
 */

import Icon from './icon/Icon'
export * from './icon/Icon'
export default Icon
