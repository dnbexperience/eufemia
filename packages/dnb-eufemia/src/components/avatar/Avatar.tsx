import React from 'react'
import classnames from 'classnames'

// Components
import { createSpacingClasses } from '../space/SpacingHelper'
import { createSkeletonClass } from '../skeleton/SkeletonHelper'

// Elements
import Img, { ImgProps } from '../../elements/img/Img'

// Shared
import Context from '../../shared/Context'
import type { SpacingProps } from '../../shared/types'
import type { SkeletonShow } from '../skeleton/Skeleton'
import {
  validateDOMAttributes,
  extendPropsWithContext,
  warn,
} from '../../shared/component-helper'

// Internal
import AvatarGroup, { AvatarGroupContext } from './AvatarGroup'

export type AvatarSizes = 'small' | 'medium' | 'large' | 'x-large'
export type AvatarVariants = 'primary' | 'secondary' | 'tertiary'

export type AvatarImgProps = ImgProps

export interface AvatarProps {
  /**
   * Used in combination with `src` to provide an alt attribute for the `img` element.
   * Default: null
   */
  alt?: string

  /**
   * Custom className on the component root
   * Default: null
   */
  className?: string

  /**
   * Skeleton should be applied when loading content
   * Default: null
   */
  skeleton?: SkeletonShow

  /**
   * The content of the component. Can be used instead of prop "data".
   * Default: null
   */
  children?: React.ReactNode

  /**
   * The size of the component.
   * Default: medium.
   */
  size?: AvatarSizes

  /**
   * Specifies the path to the image
   * Default: null
   */
  src?: string

  /**
   * Props applied to the `img` element if the component is used to display an image.
   * Default: null
   */
  imgProps?: ImgProps

  /**
   * The variant of the component.
   * Default: primary.
   */
  variant?: AvatarVariants

  /**
   * If an avatar is hidden from the screen reader (by setting aria-hidden={true}) or if label is given, typical inside a table or dl (definition list), then you can disable Avatar.Group as a dependent of Avatar.
   * Use `true` to omit the `Avatar group required:` warning.
   * Default: null
   */
  hasLabel?: boolean
}

export const defaultProps = {
  size: 'medium',
  variant: 'primary',
  skeleton: false,
}

const Avatar = (localProps: AvatarProps & SpacingProps) => {
  // Every component should have a context
  const context = React.useContext(Context)
  const avatarGroupContext = React.useContext(AvatarGroupContext)

  // Extract additional props from global context
  const allProps = extendPropsWithContext(
    localProps,
    defaultProps,
    context?.Avatar,
    { skeleton: context?.skeleton },
    avatarGroupContext
  )

  const {
    alt,
    className,
    children: childrenProp,
    size,
    skeleton,
    variant,
    src,
    imgProps,
    hasLabel,
    ...props
  } = allProps

  let children = null

  const skeletonClasses = createSkeletonClass('shape', skeleton, context)
  const spacingClasses = createSpacingClasses(props)

  const childrenIsString = typeof childrenProp === 'string'

  if (src || imgProps) {
    const imageProps = { src, alt, ...imgProps }
    children = <Img {...imageProps} />
  } else if (childrenIsString) {
    const firstLetterUpperCase = childrenProp.charAt(0).toUpperCase()
    children = <span aria-hidden>{firstLetterUpperCase}</span>
  } else {
    children = childrenProp
  }

  if (!avatarGroupContext && !hasLabel) {
    warn(
      `Avatar group required: An Avatar requires an Avatar.Group with label description as a parent component. This is to ensure correct semantic and accessibility.`
    )
  }

  validateDOMAttributes(allProps, props)

  return (
    <span
      className={classnames(
        'dnb-avatar',
        `dnb-avatar--${variant || 'primary'}`,
        `dnb-avatar--size-${size || 'medium'}`,
        skeletonClasses,
        spacingClasses,
        className
      )}
      {...props}
    >
      {childrenIsString && (
        <span className="dnb-sr-only">{childrenProp}</span>
      )}
      {children}
    </span>
  )
}

Avatar.Group = AvatarGroup

export { AvatarGroup }

Avatar._supportsSpacingProps = true

export default Avatar
