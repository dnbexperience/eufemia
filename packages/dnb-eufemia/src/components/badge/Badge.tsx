import React from 'react'
import classnames from 'classnames'

// Components
import { createSpacingClasses } from '../space/SpacingHelper'
import { createSkeletonClass } from '../skeleton/SkeletonHelper'

// Shared
import Context from '../../shared/Context'
import type { SpacingProps } from '../../shared/types'
import type { SkeletonShow } from '../skeleton/Skeleton'
import {
  warn,
  extendPropsWithContext,
  validateDOMAttributes,
} from '../../shared/component-helper'

export type BadgeProps = {
  /**
   * Aria label to describe the badge
   * Default: null
   */
  label?: React.ReactNode

  /**
   * Custom className on the component root
   * Default: null
   */
  className?: string

  /**
   * Skeleton should be applied when loading content
   * Default: null
   */
  skeleton?: SkeletonShow

  /**
   * The content to display the badge on top of.
   * Default: null
   */
  children?: React.ReactNode

  /**
   * The content of the component.
   * Default: null
   */
  content?: string | number | React.ReactNode

  /**
   * The vertical positioning of the component.
   * Default: null
   */
  vertical?: 'bottom' | 'top'

  /**
   * The horizontal positioning of the component.
   * Default: null
   */
  horizontal?: 'left' | 'right'

  /**
   * The variant of the component.
   * Default: information.
   */
  variant?: 'information' | 'notification'
}

type BadgeAndSpacingProps = BadgeProps & SpacingProps

export const defaultProps = {
  label: null,
  className: null,
  skeleton: false,
  children: null,
  content: null,
  vertical: null,
  horizontal: null,
  variant: 'information',
}

function Badge(localProps: BadgeAndSpacingProps) {
  // Every component should have a context
  const context = React.useContext(Context)

  // Extract additional props from global context
  const allProps = extendPropsWithContext(
    localProps,
    defaultProps,
    context?.Badge,
    { skeleton: context?.skeleton }
  )

  const {
    label,
    className,
    children, // eslint-disable-line
    skeleton,
    horizontal,
    vertical,
    content: contentProp,
    variant,
    ...props
  } = allProps

  validateDOMAttributes(allProps, props)

  if (children) {
    return (
      <BadgeRoot>
        {children}
        <BadgeElem />
      </BadgeRoot>
    )
  }

  return <BadgeElem />

  function BadgeRoot({ children }: { children: React.ReactNode }) {
    return <span className="dnb-badge__root">{children}</span>
  }

  function BadgeElem() {
    const skeletonClasses = createSkeletonClass('shape', skeleton, context)
    const spacingClasses = createSpacingClasses(allProps)
    const contentIsNum = typeof contentProp === 'number'
    const variantIsNotification = variant === 'notification'

    const content =
      variantIsNotification && contentIsNum && contentProp > 9
        ? '9+'
        : contentProp

    if (variantIsNotification && !contentIsNum) {
      warn(
        `Type of content should be a number: A notification badge is best suited to display content of type number.`
      )
    }
    if (!label && contentIsNum) {
      warn(
        `Label required: A Badge with a number as content requires a label describing the content of the badge. This is to ensure correct semantic and accessibility.`
      )
    }

    const isInline = !children && content

    return (
      <span
        role="status"
        className={classnames(
          'dnb-badge',
          `dnb-badge--variant-${variant}`,
          horizontal && `dnb-badge--horizontal-${horizontal}`,
          vertical && `dnb-badge--vertical-${vertical}`,
          isInline && 'dnb-badge--inline',
          skeletonClasses,
          spacingClasses,
          className
        )}
        {...props}
      >
        {label && <span className="dnb-sr-only">{label} </span>}
        {content}
      </span>
    )
  }
}

Badge._supportsSpacingProps = true

export default Badge
