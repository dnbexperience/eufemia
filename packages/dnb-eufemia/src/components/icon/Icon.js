/**
 * Web Icon Component
 *
 */

import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { ErrorHandler } from '../../shared/error-helper'
import {
  isTrue,
  validateDOMAttributes,
  processChildren,
  extendPropsWithContextInClassComponent,
} from '../../shared/component-helper'
import {
  spacingPropTypes,
  createSpacingClasses,
} from '../space/SpacingHelper'
import { createSkeletonClass } from '../skeleton/SkeletonHelper'
import Context from '../../shared/Context'
import { iconCase } from './IconHelpers'

export const DefaultIconSize = 16
export const DefaultIconSizes = {
  // small: 8, // currently not in use
  default: 16,
  medium: 24,
  // large: 32 // currently not in use
}
// instead of using Object.entries(DefaultIconSizes)
export const ListDefaultIconSizes = [
  ['default', 16],
  ['medium', 24],
]
export const ValidIconSizes = [
  'small', // 12px 0.75rem
  'default', // 16px 1rem
  'medium', // 24px 1.5rem
  'large', // 32px 2rem
  'x-large', // 40px 2.5rem
  'xx-large', // 48px 3rem
]

export const iconPropTypes = {
  icon: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.func,
  ]),
  modifier: PropTypes.string,
  size: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
    PropTypes.oneOf(['default', 'medium', 'large']),
  ]),

  ...spacingPropTypes,

  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  border: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  color: PropTypes.string,
  inherit_color: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  alt: PropTypes.string,
  title: PropTypes.string,
  skeleton: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
  attributes: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),

  className: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
}

/**
 * The icon component is a span wrapping an inline svg. When using this component in your preferred framework. To load an svg file dynamically, you may need a "svg-loader". Feel free to use whatever tool you want (regarding the setup/tooling), as long as the output is the same markup as shown below.
 */
export default class Icon extends React.PureComponent {
  static contextType = Context

  static propTypes = {
    /**
     * Use spread, so generateTypes.js makes a valid copy to create the types
     */
    ...iconPropTypes,
  }

  static defaultProps = {
    icon: null,
    modifier: null,
    size: null,
    width: null,
    height: null,
    border: null,
    color: null,
    inherit_color: true,
    alt: null,
    title: null,
    skeleton: null,
    attributes: null,

    className: null,
    children: null,
  }

  static getIcon(props) {
    if (props.icon) {
      return props.icon
    }
    return processChildren(props)
  }

  render() {
    // use only the props from context, who are available here anyway
    const props = extendPropsWithContextInClassComponent(
      this.props,
      Icon.defaultProps,
      { skeleton: this.context?.skeleton },
      this.context.Icon
    )

    const { icon, size, wrapperParams, iconParams, alt } = prepareIcon(
      props,
      this.context
    )

    if (!icon) {
      return null
    }

    const IconContainer = prerenderIcon({ icon, size, alt })

    // make sure we return an empty span if we don't could get the icon
    if (!IconContainer) return <></>

    return (
      <span {...wrapperParams}>
        <IconContainer {...iconParams} />
      </span>
    )
  }
}

export const getIconNameFromComponent = (icon) => {
  const name =
    typeof icon === 'string'
      ? icon
      : icon && (icon.displayName || icon.name)
  if (/^data:image\//.test(name)) {
    return null
  }
  return name
}

export const calcSize = (props) => {
  const { icon, size, width, height } = props

  let sizeAsInt = -1
  let sizeAsString = null

  // if there is no size, check if we can find the actual size in the name
  if (!size || size === DefaultIconSize) {
    // get the icon name - we use is for several things
    const name = getIconNameFromComponent(icon)

    const nameParts = String(name || '').split('_')

    if (nameParts.length > 1) {
      const lastPartOfIconName = nameParts.reverse()[0]
      const potentialSize = ListDefaultIconSizes.filter(
        ([key]) => key === lastPartOfIconName
      ).reduce((acc, [key, value]) => {
        return key && value
      }, null)
      if (potentialSize) {
        sizeAsInt = potentialSize
      }
      if (ValidIconSizes.includes(lastPartOfIconName)) {
        sizeAsString = lastPartOfIconName
      }
    } else {
      if (typeof icon === 'function') {
        const elem = icon()
        if (elem.props) {
          let potentialSize = -1
          if (elem.props.width) {
            potentialSize = elem.props.width
          }
          if (!potentialSize && elem.props.viewBox) {
            potentialSize = /[0-9]+ [0-9]+ ([0-9]+)/.exec(
              elem.props.viewBox
            )[1] // get the width
          }
          if (potentialSize) {
            sizeAsInt = potentialSize
          }
        }
      }
    }
  }

  // if size is defined as a string, find the size number
  else if (typeof size === 'string' && !(parseFloat(size) > 0)) {
    sizeAsInt = ListDefaultIconSizes.filter(
      ([key]) => key === size
    ).reduce((acc, [key, value]) => {
      return key && value
    }, -1)

    // of if the size is a default size defined as a string
    if (ValidIconSizes.includes(size)) {
      sizeAsString = size
    }
  }

  // check if the size is given as a number, and if is a default size
  else if (parseFloat(size) > 0) {
    sizeAsInt = ListDefaultIconSizes.filter(
      ([key, value]) => key && value === parseFloat(size)
    ).reduce((acc, [key, value]) => {
      if (key && value) return value
      return acc
    }, -1)

    // has custom size
    if (sizeAsInt === -1) {
      sizeAsInt = parseFloat(size)
      sizeAsString = 'custom-size'
    }
  }

  // check if the sizeAsInt is a default size - and no sizeAsString exists yet
  if (!sizeAsString && sizeAsInt > 0) {
    const potentialSizeAsString = ListDefaultIconSizes.reduce(
      (acc, [key, value]) => {
        if (key && value === sizeAsInt) {
          return key
        }
        return acc
      },
      null
    )

    if (potentialSizeAsString) {
      sizeAsString = potentialSizeAsString
    }
  }

  // define all the svg parameters
  const { sizeAsString: isCustomSize, params: iconParams } =
    prepareIconParams({
      sizeAsString,
      sizeAsInt,
      size,
      width,
      height,
    })
  if (isCustomSize) {
    sizeAsString = isCustomSize
  }

  if (!(sizeAsInt > 0)) {
    sizeAsInt = DefaultIconSize
  }

  if (size === 'auto') {
    iconParams.width = '100%'
    iconParams.height = '100%'
    sizeAsString = 'auto'
  }

  return {
    iconParams,
    sizeAsInt,
    sizeAsString,
  }
}

const prepareIconParams = ({ sizeAsString, ...rest }) => {
  const { size, width, height, sizeAsInt } = rest
  const params = {}

  if (!sizeAsString && !(sizeAsInt > 0) && parseFloat(size) > -1) {
    params.width = params.height = parseFloat(size)
  } else if (sizeAsString === 'custom-size') {
    params.width = params.height = parseFloat(sizeAsInt)
  }
  if (parseFloat(width) > -1) {
    sizeAsString = 'custom-size'
    params.width = parseFloat(width)
  }
  if (parseFloat(height) > -1) {
    sizeAsString = 'custom-size'
    params.height = parseFloat(height)
  }

  validateDOMAttributes({}, params)

  return { params, sizeAsString }
}

export const prepareIcon = (props, context) => {
  const {
    icon,
    size, // eslint-disable-line
    width,
    height,
    border,
    color,
    inherit_color,
    modifier,
    alt,
    title,
    skeleton,
    class: _className,
    className,
    ...attributes
  } = props

  const { sizeAsString, iconParams } = calcSize({
    icon,
    size,
    width,
    height,
  })

  if (color) {
    iconParams.color = color
  }

  const label = icon ? getIconNameFromComponent(icon) : null

  // some wrapper params
  // also used for code markup simulation
  const wrapperParams = validateDOMAttributes(props, {
    role: alt ? 'img' : 'presentation',
    alt, // in case the image don't shows up (because we define the role to be img)
    'aria-label':
      label && !label.includes('default')
        ? label.replace(/_/g, ' ') + ' icon'
        : null, // for screen readers only
    title, // to show on hover, if defined
    ...attributes,
  })
  if (!alt && typeof wrapperParams['aria-hidden'] === 'undefined') {
    wrapperParams['aria-hidden'] = true
  }
  if (wrapperParams['aria-hidden']) {
    if (
      !wrapperParams['data-testid'] &&
      typeof process !== 'undefined' &&
      process.env.NODE_ENV === 'test'
    ) {
      wrapperParams['data-testid'] = wrapperParams['aria-label']
    }
    delete wrapperParams['aria-label']
  }

  wrapperParams.className = classnames(
    'dnb-icon',
    modifier && `dnb-icon--${modifier}`,
    isTrue(border) && 'dnb-icon--border',
    isTrue(inherit_color) && 'dnb-icon--inherit-color',
    sizeAsString ? `dnb-icon--${sizeAsString}` : 'dnb-icon--default',
    createSkeletonClass(null, skeleton, context),
    createSpacingClasses(props),
    _className,
    className
  )

  let iconToRender = Icon.getIcon(props)

  if (iconToRender && typeof iconToRender.defaultProps !== 'undefined') {
    iconToRender = React.createElement(
      iconToRender,
      validateDOMAttributes(
        {},
        {
          color,
          icon,
          size,
          width,
          height,
        }
      )
    )
  }

  return {
    ...props,
    icon: iconToRender,
    alt,
    iconParams,
    wrapperParams,
  }
}

export const prerenderIcon = ({
  icon,
  size = null,
  listOfIcons = null,
  alt = null,
} = {}) => {
  if (typeof icon === 'string' && /^data:image\//.test(icon)) {
    return () => <img src={icon} alt={alt || 'no-alt'} />
  }

  if (typeof icon === 'function') {
    const elem = icon()
    if (React.isValidElement(elem)) {
      return icon
    }
    return elem
  }

  if (React.isValidElement(icon) || Array.isArray(icon)) {
    return () => icon
  }

  // for importing react component
  try {
    icon = iconCase(icon)
    if (
      size &&
      DefaultIconSizes[size] &&
      size !== 'default' &&
      !(parseFloat(size) > 0) &&
      !icon.includes(size)
    ) {
      icon = `${icon}_${size}`
    }
    const mod = (
      listOfIcons.dnbIcons ? listOfIcons.dnbIcons : listOfIcons
    )[icon]
    return mod && mod.default ? mod.default : mod
  } catch (e) {
    ErrorHandler(`Icon '${icon}' did not exist!`)
    return null
  }
}

Icon._supportsSpacingProps = true
