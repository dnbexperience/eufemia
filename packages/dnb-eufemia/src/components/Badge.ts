/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index badge to autogenerate all the components and extensions
 * Used by "prepareBadges"
 */

import Badge from './badge/Badge'
export * from './badge/Badge'
export default Badge
