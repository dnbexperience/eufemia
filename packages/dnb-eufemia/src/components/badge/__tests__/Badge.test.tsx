import React from 'react'
import { render, screen } from '@testing-library/react'
import Badge, { BadgeProps } from '../Badge'
import { confetti as Confetti } from '../../../icons'
import Icon from '../../Icon'
import Avatar from '../../Avatar'

import { loadScss, axeComponent } from '../../../core/jest/jestSetup'
import { Provider } from '../../../shared'

describe('Badge', () => {
  it('renders without properties', () => {
    const props: BadgeProps = {}
    render(<Badge {...props} />)

    expect(document.querySelector('.dnb-badge')).toBeInTheDocument()
  })

  it('renders content as text', () => {
    const string = 'A'
    render(<Badge content={string} />)

    expect(screen.queryByText(string)).toBeInTheDocument()
  })

  it('renders content as number', () => {
    const number = 1
    const label = 'Notifications:'
    render(<Badge content={number} label={label} />)

    expect(screen.queryByText(number)).toBeInTheDocument()
    expect(screen.queryByText(label)).toBeInTheDocument()
  })

  it('renders 9+ when content is a number with value greater than 9', () => {
    const number = 10
    const label = 'Notifications:'
    render(<Badge content={number} variant="notification" label={label} />)

    expect(screen.queryByText('9+')).toBeInTheDocument()
    expect(screen.queryByText(label)).toBeInTheDocument()
  })

  it('renders the label as string', () => {
    const label = 'Money:'
    const content = 100
    render(<Badge label={label} content={content} />)

    expect(screen.queryByText(label)).toBeInTheDocument()
    expect(screen.queryByText(content)).toBeInTheDocument()
  })

  it('renders the label as a react node', () => {
    const label = <span data-testid="react-node">ReactNode</span>
    render(<Badge label={label} content="something" />)

    expect(screen.queryByTestId('react-node')).toBeInTheDocument()
  })

  it('renders children as content', () => {
    render(
      <Badge content={<Icon icon={Confetti} />}>
        <Avatar.Group label="children:">
          <Avatar>A</Avatar>
        </Avatar.Group>
      </Badge>
    )

    expect(screen.queryByTestId('confetti icon')).toBeInTheDocument()
  })

  it('warns when notification badge content is a string', () => {
    process.env.NODE_ENV = 'development'
    global.console.log = jest.fn()
    render(<Badge variant="notification" content="string" />)
    expect(global.console.log).toHaveBeenCalled()
  })

  it('does not warn when notification badge content is a number', () => {
    process.env.NODE_ENV = 'development'
    global.console.log = jest.fn()
    render(
      <Badge variant="notification" content={1} label="Notifications:" />
    )
    expect(global.console.log).not.toHaveBeenCalled()
  })

  it('warns when Badge content is a number and is missing a label', () => {
    process.env.NODE_ENV = 'development'
    global.console.log = jest.fn()
    render(<Badge content={1} />)
    expect(global.console.log).toHaveBeenCalled()
  })

  it('does not warn when Badge content is a string and label is missing', () => {
    process.env.NODE_ENV = 'development'
    global.console.log = jest.fn()
    render(<Badge content="text" />)
    expect(global.console.log).not.toHaveBeenCalled()
  })

  it('does not warn when Badge content is a number and has a label', () => {
    process.env.NODE_ENV = 'development'
    global.console.log = jest.fn()
    render(<Badge content="text" label="Notifications:" />)
    expect(global.console.log).not.toHaveBeenCalled()
  })

  it('should support spacing props', () => {
    render(
      <Badge
        top="2rem"
        aria-label="Info about the badge"
        content="content"
      />
    )

    const element = document.querySelector('.dnb-badge')
    const attributes = Array.from(element.attributes).map(
      (attr) => attr.name
    )

    expect(attributes).toEqual(['role', 'class', 'aria-label'])
    expect(Array.from(element.classList)).toEqual([
      'dnb-badge',
      'dnb-badge--variant-information',
      'dnb-space__top--large',
    ])
  })

  it('should have role="status"', () => {
    render(<Badge content="content" />)

    const element = document.querySelector('.dnb-badge')

    expect(element.getAttribute('role')).toBe('status')
  })

  it('should inherit skeleton prop from provider', () => {
    render(
      <Provider skeleton>
        <Badge content="content" />
      </Provider>
    )

    const element = document.querySelector('.dnb-badge')

    expect(Array.from(element.classList)).toEqual([
      'dnb-badge',
      'dnb-badge--variant-information',
      'dnb-skeleton',
      'dnb-skeleton--shape',
    ])
  })

  describe('default values', () => {
    it('has variant information as default', () => {
      render(<Badge />)

      expect(
        document.getElementsByClassName('dnb-badge--variant-information')
      ).toHaveLength(1)
    })
  })
})

describe('Badge aria', () => {
  it('should validate', async () => {
    const Component = render(<Badge content="1" />)
    expect(await axeComponent(Component)).toHaveNoViolations()
  })
})

describe('Badge scss', () => {
  it('has to match style dependencies css', () => {
    const css = loadScss(require.resolve('../style/deps.scss'))
    expect(css).toMatchSnapshot()
  })

  it('have to match default theme snapshot', () => {
    const css = loadScss(
      require.resolve('../style/themes/dnb-badge-theme-ui.scss')
    )
    expect(css).toMatchSnapshot()
  })
})
