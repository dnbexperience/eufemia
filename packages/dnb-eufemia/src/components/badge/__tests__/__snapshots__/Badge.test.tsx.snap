// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge scss has to match style dependencies css 1`] = `
"/*
* Used for snapshot testing
*
*/
/*
 * Utilities
 */
/*
* Badge component
*
*/
/*
 * Utilities
 */
.dnb-badge {
  --badge-font-size: var(--font-size-x-small);
  --badge-width: 1.5rem;
  --badge-height: 1.5rem;
  --badge-border-radius-information: calc(var(--badge-height) / 4);
  --badge-border-radius-notification: calc(var(--badge-height) / 2);
  --badge-line-height: var(--line-height-x-small);
  display: inline-flex;
  font-weight: var(--font-weight-medium);
  flex-flow: row wrap;
  white-space: nowrap;
  justify-content: center;
  align-content: center;
  align-items: center;
  width: var(--badge-width);
  height: var(--badge-height);
  font-size: var(--badge-font-size);
}
.dnb-badge, .dnb-core-style .dnb-badge {
  line-height: var(--badge-line-height);
}
.dnb-badge__root {
  position: relative;
  display: inline-flex;
  vertical-align: middle;
  flex-shrink: 0;
}
.dnb-badge__root .dnb-badge {
  display: flex;
  overflow: hidden;
  position: absolute;
  z-index: 2;
}
.dnb-badge--variant-notification {
  border-radius: var(--badge-border-radius-notification);
}
.dnb-badge--variant-information {
  border-radius: var(--badge-border-radius-information);
  min-width: var(--badge-width);
  width: auto;
  padding: 0 var(--spacing-xx-small);
}
.dnb-badge--horizontal-left {
  left: 0;
}
.dnb-badge--horizontal-right {
  right: 0;
}
.dnb-badge--vertical-bottom {
  bottom: 0;
}
.dnb-badge--vertical-top {
  top: 0;
}
.dnb-badge--inline.dnb-badge--vertical-top {
  transform: translateY(-25%);
}
.dnb-badge--inline.dnb-badge--vertical-bottom {
  transform: translateY(25%);
}"
`;

exports[`Badge scss have to match default theme snapshot 1`] = `
"/*
* Badge theme
*
*/
/*
 * Utilities
 */
.dnb-badge {
  color: var(--color-white);
}
.dnb-badge--variant-information {
  background-color: var(--color-emerald-green);
}
.dnb-badge--variant-notification {
  background-color: var(--color-fire-red);
}"
`;
